### DISCUSSION TITLE
Routing in a straight line isn't happening

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3086] [General]
I was testing some other things and noticed that sometimes I see routing happening in a straight line without going to a lower metal layer, others it doesn't happen.

I believe what is happening is that a buffer is needed in some cases and not others, which is down to the details of the .sdc file?

Run:

```
make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk
```

Then:

```
make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk gui_final
```

Here we can see routing in a straight line without changing metal layers between the elements, but not at the edges of the array:

![image](https://user-images.githubusercontent.com/2798822/227799815-a54ace6b-93ba-4589-be3e-4d9e1e07b8f0.png)

Consider the element in the upper right corner:

![image](https://user-images.githubusercontent.com/2798822/227799862-8aa8e0f3-40ff-4ece-b601-8f7851e49e62.png)

Instead of staying on M4, the router is dropping down to M2:

![image](https://user-images.githubusercontent.com/2798822/227799898-11731f1d-7932-4f0b-9f96-45730331938a.png)

I see a buffer is inserted:

![image](https://user-images.githubusercontent.com/2798822/227800127-34acfc60-be4b-4559-88d5-32b4646f5339.png)

I believe there's some interaction with the PDN grid where sometimes horizontal routes are going all the way up to M6:

![image](https://user-images.githubusercontent.com/2798822/227800358-5455470c-93e1-47ef-b6ae-8c55dea6fbd7.png)

Sometimes vertical routes need to go all the way up to M7:

![image](https://user-images.githubusercontent.com/2798822/227800426-d2996f1a-f7a1-4713-8644-49557cd9dcac.png)

**Comment by maliberty** [comment]
buffer_ports puts buffers on all the IOs so that is expected.  You could try skipping that if you want to avoid them.   The router has to drop down to the layer where the buffer pins are.

There are pdn stripes on m5/m6 so the router will have to avoid those.  You can always customize the grid

**Reply by oharboe** [reply]
`buffer_ports` is invoked unconditionally, unless `FOOTPRINT` is defined. `FOOTPRINT` is defined for chip level macros, looking at some examples:

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/d93af61780aaa8f9de06f0db870a45e6135f3378/flow/scripts/resize.tcl#L30

**Reply by maliberty** [reply]
You can make it conditional if that's useful.

**Reply by oharboe** [reply]
I am thinking that if this performance problem has to be solved anyway, then it may be an unnecessary complication


How long should writing a .lib file take? Seconds, minutes, hours?

**Reply by maliberty** [reply]
It depends on the block size but minutes more than hours.

**Reply by oharboe** [reply]
Gotit.

I think the next step is for me to run perf on my local case. It would not surprise me if I find some sort of obvious in retrospect performance pathology. That usually happens when running perf the first time on some workload that's orders of magnitude too slow.

**Reply by oharboe** [reply]
https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/940
